typescript-rewrite/samlang-core/src/interpreter/__tests__/module-interpreter.test.ts

```diff
@@ -0,0 +1,150 @@
+    type: 'object',
+    names: ['types'],
+    mappings: {
+      types: {
+        type: intType,
+        isPublic: true,
+      },
+    },
+  },
+};
+
+const memberMainFunctionNoArgs: ClassMemberDefinition = {
+  range: new Range(new Position(1, 10), new Position(3, 4)),
+  isPublic: true,
+  isMethod: false,
+  nameRange: new Range(new Position(12, 34), new Position(34, 45)),
+  name: 'main',
+  typeParameters: ['param'],
+  type: {
+    type: 'FunctionType',
+    argumentTypes: [intType],
+    returnType: intType,
+  },
+  parameters: [],
+  body: EXPRESSION_INT(new Range(new Position(123, 45), new Position(145, 89)), BigInt(2)),
+};
+
+const memberMainMethodNoArgs: ClassMemberDefinition = {
+  ...memberMainFunctionNoArgs,
+  isMethod: true,
+};
+
+const memberMainFunctionWithArgs: ClassMemberDefinition = {
+  ...memberMainFunctionNoArgs,
+  parameters: [
+    {
+      name: 'param',
+      nameRange: new Range(new Position(231, 34), new Position(88, 78)),
+      type: intType,
+      typeRange: new Range(new Position(123, 98), new Position(124, 78)),
+    },
+  ],
+};
+
+const moduleNoMainClass: SamlangModule = {
+  imports: [],
+  classes: [exampleClassDef],
+};
+
+const moduleWithMainClassNoMainFunction: SamlangModule = {
+  imports: [],
+  classes: [mainClassDef],
+};
+
+const moduleWithMainClassAndMainFunctionNoArgs: SamlangModule = {
+  imports: [],
+  classes: [
+    {
+      ...mainClassDef,
+      members: [memberMainFunctionNoArgs],
+    },
+  ],
+};
+
+const moduleWithMainClassAndMainMethodNoArgs: SamlangModule = {
+  imports: [],
+  classes: [
+    {
+      ...mainClassDef,
+      members: [memberMainMethodNoArgs],
+    },
+  ],
+};
+
+const moduleWithMainClassAndMainFunctionWithArgs: SamlangModule = {
+  imports: [],
+  classes: [
+    {
+      ...mainClassDef,
+      members: [memberMainFunctionWithArgs],
+    },
+  ],
+};
+
+it('module evaluates correctly', () => {
+  expect(moduleInterpreter.eval(moduleEmpty)).toEqual({ type: 'unit' });
+  expect(moduleInterpreter.eval(moduleNoMainClass)).toEqual({ type: 'unit' });
+  expect(moduleInterpreter.eval(moduleWithMainClassNoMainFunction)).toEqual({ type: 'unit' });
+  expect(moduleInterpreter.eval(moduleWithMainClassAndMainFunctionNoArgs)).toEqual(BigInt(2));
+  expect(moduleInterpreter.eval(moduleWithMainClassAndMainMethodNoArgs)).toEqual({ type: 'unit' });
+  expect(moduleInterpreter.eval(moduleWithMainClassAndMainFunctionWithArgs)).toEqual({
+    type: 'unit',
+  });
+});
+
+it('module runs correctly', () => {
+  expect(moduleInterpreter.run(moduleEmpty)).toEqual('');
+  expect(moduleInterpreter.run(moduleNoMainClass)).toEqual('');
+  expect(moduleInterpreter.run(moduleWithMainClassNoMainFunction)).toEqual('');
+  expect(moduleInterpreter.run(moduleWithMainClassAndMainFunctionNoArgs)).toEqual('');
+  expect(moduleInterpreter.run(moduleWithMainClassAndMainMethodNoArgs)).toEqual('');
+  expect(moduleInterpreter.run(moduleWithMainClassAndMainFunctionWithArgs)).toEqual('');
```

Comment:
Ideally, we should have some tests where the program actually print out some stuff

Reply:
hmm ok I'll add some tests using `println`

Reply:
I also realized I never tested `println` properly in `expression-interpreter` so I added that too 😅